From 5c1ebc674fea827d20a6caaadef051c775a3f251 Mon Sep 17 00:00:00 2001
From: Scott Bender <scott@scottbender.net>
Date: Wed, 28 May 2025 13:12:56 -0400
Subject: [PATCH 1/2] fix: ignore static self data

---
 src/index.ts | 57 +++++++++++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 54 insertions(+), 3 deletions(-)

diff --git a/src/index.ts b/src/index.ts
index 50993ec37..6ac11cd64 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -315,11 +315,13 @@ class Server {
         if (data.updates.length < 1) return
 
         try {
-          const preferredDelta = toPreferredDelta(data, now, app.selfContext)
+          let delta = filterStaticSelfData(data, app.selfContext)
+          delta = toPreferredDelta(delta, now, app.selfContext)
+
           if (skVersion == SKVersion.v1) {
-            deltachainV1.process(preferredDelta)
+            deltachainV1.process(delta)
           } else {
-            deltachainV2.process(preferredDelta)
+            deltachainV2.process(delta)
           }
         } catch (err) {
           console.error(err)
@@ -673,3 +675,52 @@ async function startInterfaces(
     })
   )
 }
+
+function filterStaticSelfData(delta: any, selfContext: string) {
+  if (delta.context === selfContext) {
+    delta.updates &&
+      delta.updates.forEach((update: any) => {
+        if ('values' in update) {
+          update.values = update.values.reduce((acc: any, pathValue: any) => {
+            const nvp = filterSelfDataKP(pathValue)
+            if (nvp) {
+              acc.push(nvp)
+            }
+            return acc
+          }, [])
+          if (update.values.length == 0) {
+            delete update.values
+          }
+        }
+      })
+  }
+  return delta
+}
+
+function filterSelfDataKP(pathValue: any) {
+  const emptyKeys = ['name', 'mmsi']
+
+  const filteredPaths = [
+    'design.aisShipType',
+    'design.beam',
+    'design.length',
+    'design.draft',
+    'sensors.gps.fromBow',
+    'sensors.gps.fromCenter'
+  ]
+
+  if (pathValue.path === '') {
+    if (Object.keys(pathValue.value).some((k) => emptyKeys.includes(k))) {
+      const value = pathValue.value
+      pathValue.value = {}
+      Object.keys(value).forEach((k) => {
+        if (!emptyKeys.includes(k)) {
+          pathValue.value[k] = value[k]
+        }
+      })
+    }
+  } else if (filteredPaths.includes(pathValue.path)) {
+    return null
+  }
+  return pathValue
+}

From 462c6edfe72c70ab6df51045fdd18a18dded8831 Mon Sep 17 00:00:00 2001
From: Scott Bender <scott@scottbender.net>
Date: Wed, 28 May 2025 13:36:03 -0400
Subject: [PATCH 2/2] fix: allow messages from defaults through

---
 src/index.ts       | 2 +-
 test/deltacache.js | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/index.ts b/src/index.ts
index 6ac11cd64..90cf6004c 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -680,7 +680,7 @@ function filterStaticSelfData(delta: any, selfContext: string) {
   if (delta.context === selfContext) {
     delta.updates &&
       delta.updates.forEach((update: any) => {
-        if ('values' in update) {
+        if ('values' in update && update['$source'] !== 'defaults') {
           update.values = update.values.reduce((acc: any, pathValue: any) => {
             const nvp = filterSelfDataKP(pathValue)
             if (nvp) {
diff --git a/test/deltacache.js b/test/deltacache.js
index abfe21978..358df64b0 100644
--- a/test/deltacache.js
+++ b/test/deltacache.js
@@ -46,6 +46,7 @@ const testDelta = {
     },
     {
       timestamp: '2014-05-03T09:14:11.096Z',
+      $source: 'defaults',
       values: [
         {
           path: '',
