From 06e3d19729b49484b2ad3e2832db3344361ece5c Mon Sep 17 00:00:00 2001
From: Karl-Erik Gustafsson <ke.gustafsson@gmail.com>
Date: Sun, 1 Sep 2024 12:23:13 +0300
Subject: [PATCH 1/4] fix null

---
 src/tokensecurity.js | 40 ++++++++++++++++++++++------------------
 1 file changed, 22 insertions(+), 18 deletions(-)

diff --git a/src/tokensecurity.js b/src/tokensecurity.js
index 5220d941d..7d6a93cb7 100644
--- a/src/tokensecurity.js
+++ b/src/tokensecurity.js
@@ -557,27 +557,31 @@ module.exports = function (app, config) {
         return (
           (update.values &&
             update.values.find((valuePath) => {
-              return (
-                strategy.checkACL(
-                  req.skPrincipal.identifier,
-                  context,
-                  valuePath.path,
-                  source,
-                  'write'
-                ) === false
-              )
+              if (valuePath != null) {
+                return (
+                  strategy.checkACL(
+                    req.skPrincipal.identifier,
+                    context,
+                    valuePath.path,
+                    source,
+                    'write'
+                  ) === false
+                )
+              }
             })) ||
           (update.meta &&
             update.meta.find((valuePath) => {
-              return (
-                strategy.checkACL(
-                  req.skPrincipal.identifier,
-                  context,
-                  valuePath.path,
-                  source,
-                  'write'
-                ) === false
-              )
+              if (valuePath != null) {
+                return (
+                  strategy.checkACL(
+                    req.skPrincipal.identifier,
+                    context,
+                    valuePath.path,
+                    source,
+                    'write'
+                  ) === false
+                )
+              }
             }))
         )
       })

From f734959caec7be93dad3664b22cc9b19bbb73768 Mon Sep 17 00:00:00 2001
From: Karl-Erik Gustafsson <ke.gustafsson@gmail.com>
Date: Sun, 1 Sep 2024 12:24:53 +0300
Subject: [PATCH 2/4] fix null

---
 src/interfaces/ws.js | 37 +++++++++++++++++++------------------
 1 file changed, 19 insertions(+), 18 deletions(-)

diff --git a/src/interfaces/ws.js b/src/interfaces/ws.js
index ed02b3c40..755a75db5 100644
--- a/src/interfaces/ws.js
+++ b/src/interfaces/ws.js
@@ -504,26 +504,27 @@ function processUpdates(app, pathSources, spark, msg) {
 
       if (source) {
         update.values.forEach((valuePath) => {
-          if (!pathSources[valuePath.path]) {
-            pathSources[valuePath.path] = {}
-          }
-          if (
-            !pathSources[valuePath.path][source] ||
-            pathSources[valuePath.path][source] !== spark
-          ) {
-            if (pathSources[valuePath.path][source]) {
-              console.log(
-                `WARNING: got a new ws client for path ${valuePath.path} source ${source}`
+          if (valuePath != null) {
+            if (!pathSources[valuePath.path]) {
+              pathSources[valuePath.path] = {}
+            }
+            if (
+              !pathSources[valuePath.path][source] ||
+              pathSources[valuePath.path][source] !== spark
+            ) {
+              if (pathSources[valuePath.path][source]) {
+                console.log(
+                  `WARNING: got a new ws client for path ${valuePath.path} source ${source}`
+                )
+              }
+              debug(
+                'registered spark for source %s path %s = %s',
+                source,
+                valuePath.path,
+                spark.id
               )
+              pathSources[valuePath.path][source] = spark
             }
-            debug(
-              'registered spark for source %s path %s = %s',
-              source,
-              valuePath.path,
-              spark.id
-            )
-
-            pathSources[valuePath.path][source] = spark
           }
         })
       }

From e3a8bfa29b7e30d4bbfce199ff9b6bbcff15f00b Mon Sep 17 00:00:00 2001
From: Karl-Erik Gustafsson <ke.gustafsson@gmail.com>
Date: Sun, 1 Sep 2024 12:26:41 +0300
Subject: [PATCH 3/4] fix null

---
 src/deltaPriority.ts | 32 +++++++++++++++++---------------
 1 file changed, 17 insertions(+), 15 deletions(-)

diff --git a/src/deltaPriority.ts b/src/deltaPriority.ts
index cb0b0e682..89695a3eb 100644
--- a/src/deltaPriority.ts
+++ b/src/deltaPriority.ts
@@ -149,27 +149,29 @@ export const getToPreferredDelta = (
           update.values =
             update.values &&
             update.values.reduce((acc: any, pathValue: PathValue) => {
-              const latest = getLatest(
-                delta.context as Context,
-                pathValue.path as Path
-              )
-              const isPreferred = isPreferredValue(
-                pathValue.path as Path,
-                latest,
-                update.$source,
-                millis
-              )
-              if (isPreferred) {
-                setLatest(
+              if (pathValue != null) {
+                const latest = getLatest(
                   delta.context as Context,
+                  pathValue.path as Path
+                )
+                const isPreferred = isPreferredValue(
                   pathValue.path as Path,
-                  update.$source as SourceRef,
+                  latest,
+                  update.$source,
                   millis
                 )
-                acc.push(pathValue)
+                if (isPreferred) {
+                  setLatest(
+                    delta.context as Context,
+                    pathValue.path as Path,
+                    update.$source as SourceRef,
+                    millis
+                  )
+                  acc.push(pathValue)
+                  return acc
+                }
                 return acc
               }
-              return acc
             }, [])
         })
     }
