diff --git a/eslint.config.js b/eslint.config.js
index 8437dad..87fe65d 100644
--- a/eslint.config.js
+++ b/eslint.config.js
@@ -7,7 +7,7 @@ const react = require('eslint-plugin-react')
 const chai = require('eslint-plugin-chai-friendly')

 module.exports = defineConfig([
-  globalIgnores(['**/public', '**/dist']),
+  globalIgnores(['**/public', '**/dist', '**/.__mf__temp']),

   // TypeScript options
   {
