diff --git a/src/interfaces/ws.js b/src/interfaces/ws.js
index e7854da..222ef14 100644
--- a/src/interfaces/ws.js
+++ b/src/interfaces/ws.js
@@ -380,7 +380,7 @@ module.exports = function (app) {
         requestId: msg.requestId,
         state: 'COMPLETED',
         statusCode: 400,
-        message: 'A request has already beem submitted'
+        message: 'A request has already been submitted'
       })
     } else {
       requestAccess(
@@ -504,26 +504,28 @@ function processUpdates(app, pathSources, spark, msg) {

       if (source) {
         update.values.forEach((valuePath) => {
-          if (!pathSources[valuePath.path]) {
-            pathSources[valuePath.path] = {}
-          }
-          if (
-            !pathSources[valuePath.path][source] ||
-            pathSources[valuePath.path][source] !== spark
-          ) {
-            if (pathSources[valuePath.path][source]) {
-              console.log(
-                `WARNING: got a new ws client for path ${valuePath.path} source ${source}`
-              )
+          if (valuePath != null) {
+            if (!pathSources[valuePath.path]) {
+              pathSources[valuePath.path] = {}
             }
-            debug(
-              'registered spark for source %s path %s = %s',
-              source,
-              valuePath.path,
-              spark.id
-            )
+            if (
+              !pathSources[valuePath.path][source] ||
+              pathSources[valuePath.path][source] !== spark
+            ) {
+              if (pathSources[valuePath.path][source]) {
+                console.log(
+                  `WARNING: got a new ws client for path ${valuePath.path} source ${source}`
+                )
+              }
+              debug(
+                'registered spark for source %s path %s = %s',
+                source,
+                valuePath.path,
+                spark.id
+              )

-            pathSources[valuePath.path][source] = spark
+              pathSources[valuePath.path][source] = spark
+            }
           }
         })
       }
